// @ant-design/pro-layout/lib/components/SettingDrawer/BlockCheckbox.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd/es/tooltip/style");

var _tooltip = _interopRequireDefault(require("antd/es/tooltip"));

var _icons = require("@ant-design/icons");

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireWildcard(require("react"));

var BlockCheckbox = function BlockCheckbox(_ref) {
  var value = _ref.value,
      configType = _ref.configType,
      onChange = _ref.onChange,
      list = _ref.list,
      prefixCls = _ref.prefixCls;
  var baseClassName = "".concat(prefixCls, "-drawer-block-checkbox");
  var dom = (0, _react.useMemo)(function () {
    var domList = (list || []).map(function (item) {
      return /*#__PURE__*/_react.default.createElement(_tooltip.default, {
        title: item.title,
        key: item.key
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: (0, _classnames.default)("".concat(baseClassName, "-item"), "".concat(baseClassName, "-item-").concat(item.key), "".concat(baseClassName, "-").concat(configType, "-item")),
        onClick: function onClick() {
          return onChange(item.key);
        }
      }, /*#__PURE__*/_react.default.createElement(_icons.CheckOutlined, {
        className: "".concat(baseClassName, "-selectIcon"),
        style: {
          display: value === item.key ? 'block' : 'none'
        }
      })));
    });
    return domList; // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [value, list === null || list === void 0 ? void 0 : list.length, onChange]);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: baseClassName,
    style: {
      minHeight: 42
    }
  }, dom);
};

var _default = BlockCheckbox;
exports.default = _default;
