// @ant-design/pro-form/lib/layouts/LoginFormPage/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LoginFormPage = LoginFormPage;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

require("antd/es/config-provider/style");

var _configProvider = _interopRequireDefault(require("antd/es/config-provider"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _proProvider = require("@ant-design/pro-provider");

var _react = _interopRequireWildcard(require("react"));

var _ProForm = require("../ProForm");

require("./index.less");

var _excluded = ["logo", "message", "style", "activityConfig", "backgroundImageUrl", "title", "subTitle", "actions", "children"];

function LoginFormPage(props) {
  var logo = props.logo,
      message = props.message,
      style = props.style,
      _props$activityConfig = props.activityConfig,
      activityConfig = _props$activityConfig === void 0 ? {} : _props$activityConfig,
      backgroundImageUrl = props.backgroundImageUrl,
      title = props.title,
      subTitle = props.subTitle,
      actions = props.actions,
      children = props.children,
      proFormProps = (0, _objectWithoutProperties2.default)(props, _excluded);
  var intl = (0, _proProvider.useIntl)();

  var genSubmitButtonProps = function genSubmitButtonProps() {
    var _proFormProps$submitt, _proFormProps$submitt2;

    if (proFormProps.submitter === false) return false;
    if (((_proFormProps$submitt = proFormProps.submitter) === null || _proFormProps$submitt === void 0 ? void 0 : _proFormProps$submitt.submitButtonProps) === false) return false;
    return (0, _objectSpread2.default)({
      size: 'large',
      style: {
        width: '100%'
      }
    }, (_proFormProps$submitt2 = proFormProps.submitter) === null || _proFormProps$submitt2 === void 0 ? void 0 : _proFormProps$submitt2.submitButtonProps);
  };

  var submitter = (0, _objectSpread2.default)((0, _objectSpread2.default)({
    searchConfig: {
      submitText: intl.getMessage('loginForm.submitText', '登录')
    },
    render: function render(_, dom) {
      return dom.pop();
    }
  }, proFormProps.submitter), {}, {
    submitButtonProps: genSubmitButtonProps()
  });
  var context = (0, _react.useContext)(_configProvider.default.ConfigContext);
  var baseClassName = context.getPrefixCls('pro-form-login-page');

  var getCls = function getCls(className) {
    return "".concat(baseClassName, "-").concat(className);
  };
  /** 生成logo 的dom，如果是string 设置为图片 如果是个 dom 就原样保留 */


  var logoDom = (0, _react.useMemo)(function () {
    if (!logo) return null;

    if (typeof logo === 'string') {
      return /*#__PURE__*/_react.default.createElement("img", {
        src: logo
      });
    }

    return logo;
  }, [logo]);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: baseClassName,
    style: (0, _objectSpread2.default)((0, _objectSpread2.default)({}, style), {}, {
      backgroundImage: "url(\"".concat(backgroundImageUrl, "\")")
    })
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: getCls('notice')
  }, activityConfig && /*#__PURE__*/_react.default.createElement("div", {
    className: getCls('notice-activity'),
    style: activityConfig.style
  }, activityConfig.title && /*#__PURE__*/_react.default.createElement("div", {
    className: getCls('notice-activity-title')
  }, " ", activityConfig.title, " "), activityConfig.subTitle && /*#__PURE__*/_react.default.createElement("div", {
    className: getCls('notice-activity-subTitle')
  }, " ", activityConfig.subTitle, " "), activityConfig.action && /*#__PURE__*/_react.default.createElement("div", {
    className: getCls('notice-activity-action')
  }, " ", activityConfig.action, " "))), /*#__PURE__*/_react.default.createElement("div", {
    className: getCls('container')
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: getCls('top')
  }, title || logoDom ? /*#__PURE__*/_react.default.createElement("div", {
    className: getCls('header')
  }, logoDom ? /*#__PURE__*/_react.default.createElement("span", {
    className: getCls('logo')
  }, logoDom) : null, title ? /*#__PURE__*/_react.default.createElement("span", {
    className: getCls('title')
  }, title) : null) : null, subTitle ? /*#__PURE__*/_react.default.createElement("div", {
    className: getCls('desc')
  }, subTitle) : null), /*#__PURE__*/_react.default.createElement("div", {
    className: getCls('main')
  }, /*#__PURE__*/_react.default.createElement(_ProForm.ProForm, (0, _extends2.default)({
    isKeyPressSubmit: true
  }, proFormProps, {
    submitter: submitter
  }), message, children), actions ? /*#__PURE__*/_react.default.createElement("div", {
    className: getCls('other')
  }, actions) : null)));
}
