// @ant-design/pro-table/es/components/Dropdown/index.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import "antd/es/dropdown/style";
import _Dropdown from "antd/es/dropdown";
import "antd/es/button/style";
import _Button from "antd/es/button";
import "antd/es/menu/style";
import _Menu from "antd/es/menu";
import "antd/es/config-provider/style";
import _ConfigProvider from "antd/es/config-provider";
var _excluded = ["key", "name"];
import { DownOutlined, EllipsisOutlined } from '@ant-design/icons';
import classnames from 'classnames';
import React, { useContext } from 'react';
import './index.less';
/**
 * 一个简单的下拉菜单
 *
 * @param param0
 */

var DropdownButton = function DropdownButton(_ref) {
  var children = _ref.children,
      menus = _ref.menus,
      onSelect = _ref.onSelect,
      className = _ref.className,
      style = _ref.style;

  var _useContext = useContext(_ConfigProvider.ConfigContext),
      getPrefixCls = _useContext.getPrefixCls;

  var tempClassName = getPrefixCls('pro-table-dropdown');
  var menu = /*#__PURE__*/React.createElement(_Menu, {
    onClick: function onClick(params) {
      return onSelect && onSelect(params.key);
    }
  }, menus === null || menus === void 0 ? void 0 : menus.map(function (item) {
    return /*#__PURE__*/React.createElement(_Menu.Item, {
      key: item.key
    }, item.name);
  }));
  return /*#__PURE__*/React.createElement(_Dropdown, {
    overlay: menu,
    className: classnames(tempClassName, className)
  }, /*#__PURE__*/React.createElement(_Button, {
    style: style
  }, children, " ", /*#__PURE__*/React.createElement(DownOutlined, null)));
};

var TableDropdown = function TableDropdown(_ref2) {
  var propsClassName = _ref2.className,
      style = _ref2.style,
      onSelect = _ref2.onSelect,
      _ref2$menus = _ref2.menus,
      menus = _ref2$menus === void 0 ? [] : _ref2$menus,
      children = _ref2.children;

  var _useContext2 = useContext(_ConfigProvider.ConfigContext),
      getPrefixCls = _useContext2.getPrefixCls;

  var className = getPrefixCls('pro-table-dropdown');
  var menu = /*#__PURE__*/React.createElement(_Menu, {
    onClick: function onClick(params) {
      onSelect === null || onSelect === void 0 ? void 0 : onSelect(params.key);
    },
    items: menus.map(function (_ref3) {
      var key = _ref3.key,
          name = _ref3.name,
          rest = _objectWithoutProperties(_ref3, _excluded);

      return _objectSpread(_objectSpread({
        key: key
      }, rest), {}, {
        title: rest.title,
        label: name
      });
    })
  });
  return /*#__PURE__*/React.createElement(_Dropdown, {
    overlay: menu,
    className: classnames(className, propsClassName)
  }, /*#__PURE__*/React.createElement("a", {
    style: style
  }, children || /*#__PURE__*/React.createElement(EllipsisOutlined, null)));
};

TableDropdown.Button = DropdownButton;
export default TableDropdown;
