// @ant-design/pro-form/lib/components/Slider/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd/es/slider/style");

var _slider = _interopRequireDefault(require("antd/es/slider"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _createField = require("../../BaseForm/createField");

/**
 * 文本选择组件
 *
 * @param
 */
var BaseProFormSlider = function BaseProFormSlider(_ref, ref) {
  var fieldProps = _ref.fieldProps,
      min = _ref.min,
      max = _ref.max,
      step = _ref.step,
      marks = _ref.marks,
      vertical = _ref.vertical,
      range = _ref.range;
  return /*#__PURE__*/_react.default.createElement(_slider.default, (0, _extends2.default)({
    min: min,
    max: max,
    step: step,
    marks: marks,
    vertical: vertical,
    range: range
  }, fieldProps, {
    ref: ref
  }));
};

var ProFormSlider = (0, _createField.createField)( /*#__PURE__*/_react.default.forwardRef(BaseProFormSlider), {
  lightFilterLabelFormatter: function lightFilterLabelFormatter(value) {
    if (Array.isArray(value)) {
      return value.join('~');
    }

    return value;
  }
});
var _default = ProFormSlider;
exports.default = _default;
