// @ant-design/pro-layout/lib/components/GlobalFooter/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd/es/config-provider/style");

var _configProvider = _interopRequireDefault(require("antd/es/config-provider"));

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireWildcard(require("react"));

require("./index.less");

var _default = function _default(_ref) {
  var className = _ref.className,
      prefixCls = _ref.prefixCls,
      links = _ref.links,
      copyright = _ref.copyright,
      style = _ref.style;
  var context = (0, _react.useContext)(_configProvider.default.ConfigContext);
  var baseClassName = context.getPrefixCls(prefixCls || 'pro-global-footer');

  if ((links == null || links === false || Array.isArray(links) && links.length === 0) && (copyright == null || copyright === false)) {
    return null;
  }

  var clsString = (0, _classnames.default)(baseClassName, className);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: clsString,
    style: style
  }, links && /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(baseClassName, "-links")
  }, links.map(function (link) {
    return /*#__PURE__*/_react.default.createElement("a", {
      key: link.key,
      title: link.key,
      target: link.blankTarget ? '_blank' : '_self',
      href: link.href,
      rel: "noreferrer"
    }, link.title);
  })), copyright && /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(baseClassName, "-copyright")
  }, copyright));
};

exports.default = _default;
