// @ant-design/pro-form/lib/layouts/QueryFilter/Actions.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd/es/space/style");

var _space = _interopRequireDefault(require("antd/es/space"));

require("antd/es/config-provider/style");

var _configProvider = _interopRequireDefault(require("antd/es/config-provider"));

var _icons = require("@ant-design/icons");

var _proProvider = require("@ant-design/pro-provider");

var _proUtils = require("@ant-design/pro-utils");

var _react = _interopRequireWildcard(require("react"));

var defaultCollapseRender = function defaultCollapseRender(collapsed, _, intl, hiddenNum) {
  if (collapsed) {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, intl.getMessage('tableForm.collapsed', '展开'), hiddenNum && "(".concat(hiddenNum, ")"), /*#__PURE__*/_react.default.createElement(_icons.DownOutlined, {
      style: {
        marginLeft: '0.5em',
        transition: '0.3s all',
        transform: "rotate(".concat(collapsed ? 0 : 0.5, "turn)")
      }
    }));
  }

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, intl.getMessage('tableForm.expand', '收起'), /*#__PURE__*/_react.default.createElement(_icons.DownOutlined, {
    style: {
      marginLeft: '0.5em',
      transition: '0.3s all',
      transform: "rotate(".concat(collapsed ? 0 : 0.5, "turn)")
    }
  }));
};
/**
 * FormFooter 的组件，可以自动进行一些配置
 *
 * @param props
 */


var Actions = function Actions(props) {
  var setCollapsed = props.setCollapsed,
      _props$collapsed = props.collapsed,
      collapsed = _props$collapsed === void 0 ? false : _props$collapsed,
      submitter = props.submitter,
      style = props.style,
      hiddenNum = props.hiddenNum;

  var _useContext = (0, _react.useContext)(_configProvider.default.ConfigContext),
      getPrefixCls = _useContext.getPrefixCls;

  var intl = (0, _proProvider.useIntl)();
  var collapseRender = (0, _proUtils.omitBoolean)(props.collapseRender) || defaultCollapseRender;
  return /*#__PURE__*/_react.default.createElement(_space.default, {
    style: style,
    size: 16
  }, submitter, props.collapseRender !== false && /*#__PURE__*/_react.default.createElement("a", {
    className: getPrefixCls('pro-form-collapse-button'),
    onClick: function onClick() {
      return setCollapsed(!collapsed);
    }
  }, collapseRender === null || collapseRender === void 0 ? void 0 : collapseRender(collapsed, props, intl, hiddenNum)));
};

var _default = Actions;
exports.default = _default;
