// @ant-design/pro-list/es/Item.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _extends from "@babel/runtime/helpers/esm/extends";
import "antd/es/skeleton/style";
import _Skeleton from "antd/es/skeleton";
import "antd/es/avatar/style";
import _Avatar from "antd/es/avatar";
import "antd/es/list/style";
import _List from "antd/es/list";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import "antd/es/config-provider/style";
import _ConfigProvider from "antd/es/config-provider";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
var _excluded = ["title", "subTitle", "content", "itemTitleRender", "prefixCls", "actions", "item", "recordKey", "avatar", "cardProps", "description", "isEditable", "checkbox", "index", "selected", "loading", "expand", "onExpand", "expandable", "rowSupportExpand", "showActions", "showExtra", "type", "style", "className", "record", "onRow", "onItem", "itemHeaderRender", "cardActionProps", "extra"];
import { RightOutlined } from '@ant-design/icons';
import ProCard from '@ant-design/pro-card';
import classNames from 'classnames';
import useMergedState from "rc-util/es/hooks/useMergedState";
import React, { useContext, useMemo } from 'react';
export function renderExpandIcon(_ref) {
  var _classNames;

  var prefixCls = _ref.prefixCls,
      _ref$expandIcon = _ref.expandIcon,
      expandIcon = _ref$expandIcon === void 0 ? /*#__PURE__*/React.createElement(RightOutlined, null) : _ref$expandIcon,
      onExpand = _ref.onExpand,
      expanded = _ref.expanded,
      record = _ref.record;
  var icon = expandIcon;
  var expandClassName = "".concat(prefixCls, "-row-expand-icon");

  var onClick = function onClick(event) {
    onExpand(!expanded);
    event.stopPropagation();
  };

  if (typeof expandIcon === 'function') {
    icon = expandIcon({
      expanded: expanded,
      onExpand: onExpand,
      record: record
    });
  }

  return /*#__PURE__*/React.createElement("span", {
    className: classNames(expandClassName, (_classNames = {}, _defineProperty(_classNames, "".concat(prefixCls, "-row-expanded"), expanded), _defineProperty(_classNames, "".concat(prefixCls, "-row-collapsed"), !expanded), _classNames)),
    onClick: onClick
  }, icon);
}

function ProListItem(props) {
  var _classNames2, _ref3, _classNames4, _ref4, _classNames6;

  var customizePrefixCls = props.prefixCls;

  var _useContext = useContext(_ConfigProvider.ConfigContext),
      getPrefixCls = _useContext.getPrefixCls;

  var prefixCls = getPrefixCls('pro-list', customizePrefixCls);
  var defaultClassName = "".concat(prefixCls, "-row");

  var title = props.title,
      subTitle = props.subTitle,
      content = props.content,
      itemTitleRender = props.itemTitleRender,
      restPrefixCls = props.prefixCls,
      actions = props.actions,
      item = props.item,
      recordKey = props.recordKey,
      avatar = props.avatar,
      cardProps = props.cardProps,
      description = props.description,
      isEditable = props.isEditable,
      checkbox = props.checkbox,
      index = props.index,
      selected = props.selected,
      loading = props.loading,
      propsExpand = props.expand,
      propsOnExpand = props.onExpand,
      expandableConfig = props.expandable,
      rowSupportExpand = props.rowSupportExpand,
      showActions = props.showActions,
      showExtra = props.showExtra,
      type = props.type,
      style = props.style,
      _props$className = props.className,
      propsClassName = _props$className === void 0 ? defaultClassName : _props$className,
      record = props.record,
      onRow = props.onRow,
      onItem = props.onItem,
      itemHeaderRender = props.itemHeaderRender,
      cardActionProps = props.cardActionProps,
      extra = props.extra,
      rest = _objectWithoutProperties(props, _excluded);

  var _ref2 = expandableConfig || {},
      expandedRowRender = _ref2.expandedRowRender,
      expandIcon = _ref2.expandIcon,
      expandRowByClick = _ref2.expandRowByClick,
      _ref2$indentSize = _ref2.indentSize,
      indentSize = _ref2$indentSize === void 0 ? 8 : _ref2$indentSize,
      expandedRowClassName = _ref2.expandedRowClassName;

  var _useMergedState = useMergedState(!!propsExpand, {
    value: propsExpand,
    onChange: propsOnExpand
  }),
      _useMergedState2 = _slicedToArray(_useMergedState, 2),
      expanded = _useMergedState2[0],
      onExpand = _useMergedState2[1];

  var className = classNames((_classNames2 = {}, _defineProperty(_classNames2, "".concat(defaultClassName, "-selected"), !cardProps && selected), _defineProperty(_classNames2, "".concat(defaultClassName, "-show-action-hover"), showActions === 'hover'), _defineProperty(_classNames2, "".concat(defaultClassName, "-type-").concat(type), !!type), _defineProperty(_classNames2, "".concat(defaultClassName, "-editable"), isEditable), _defineProperty(_classNames2, "".concat(defaultClassName, "-show-extra-hover"), showExtra === 'hover'), _classNames2), defaultClassName);
  var extraClassName = classNames(_defineProperty({}, "".concat(propsClassName, "-extra"), showExtra === 'hover'));
  var needExpanded = expanded || Object.values(expandableConfig || {}).length === 0;
  var expandedRowDom = expandedRowRender && expandedRowRender(record, index, indentSize, expanded);
  var extraDom = useMemo(function () {
    if (!actions || cardActionProps === 'actions') {
      return undefined;
    }

    return [/*#__PURE__*/React.createElement("div", {
      key: "action",
      onClick: function onClick(e) {
        return e.stopPropagation();
      }
    }, actions)];
  }, [actions, cardActionProps]);
  var actionsDom = useMemo(function () {
    if (!actions || !cardActionProps || cardActionProps === 'extra') {
      return undefined;
    }

    return [/*#__PURE__*/React.createElement("div", {
      key: "action",
      onClick: function onClick(e) {
        return e.stopPropagation();
      }
    }, actions)];
  }, [actions, cardActionProps]);
  var titleDom = title || subTitle ? /*#__PURE__*/React.createElement("div", {
    className: "".concat(className, "-header-title")
  }, title && /*#__PURE__*/React.createElement("div", {
    className: "".concat(className, "-title")
  }, title), subTitle && /*#__PURE__*/React.createElement("div", {
    className: "".concat(className, "-subTitle")
  }, subTitle)) : null;
  var metaTitle = (_ref3 = itemTitleRender && (itemTitleRender === null || itemTitleRender === void 0 ? void 0 : itemTitleRender(record, index, titleDom))) !== null && _ref3 !== void 0 ? _ref3 : titleDom;
  var metaDom = metaTitle || avatar || subTitle || description ? /*#__PURE__*/React.createElement(_List.Item.Meta, {
    avatar: avatar,
    title: metaTitle,
    description: description && needExpanded && /*#__PURE__*/React.createElement("div", {
      className: "".concat(className, "-description")
    }, description)
  }) : null;
  var rowClassName = classNames((_classNames4 = {}, _defineProperty(_classNames4, "".concat(className, "-item-has-checkbox"), checkbox), _defineProperty(_classNames4, "".concat(className, "-item-has-avatar"), avatar), _defineProperty(_classNames4, className, className), _classNames4));
  var cardTitleDom = useMemo(function () {
    if (avatar || title) {
      return /*#__PURE__*/React.createElement(React.Fragment, null, avatar && /*#__PURE__*/React.createElement(_Avatar, {
        size: 22,
        src: avatar,
        className: getPrefixCls('list-item-meta-avatar')
      }), /*#__PURE__*/React.createElement("span", {
        className: getPrefixCls('list-item-meta-title')
      }, title));
    }

    return null;
  }, [avatar, getPrefixCls, title]);
  var defaultDom = !cardProps ? /*#__PURE__*/React.createElement(_List.Item, _extends({
    className: classNames(rowClassName, _defineProperty({}, propsClassName, propsClassName !== defaultClassName))
  }, rest, {
    actions: extraDom,
    extra: !!extra && /*#__PURE__*/React.createElement("div", {
      className: extraClassName
    }, extra)
  }, onRow === null || onRow === void 0 ? void 0 : onRow(record, index), onItem === null || onItem === void 0 ? void 0 : onItem(record, index), {
    onClick: function onClick(e) {
      var _onRow, _onRow$onClick, _onItem, _onItem$onClick;

      onRow === null || onRow === void 0 ? void 0 : (_onRow = onRow(record, index)) === null || _onRow === void 0 ? void 0 : (_onRow$onClick = _onRow.onClick) === null || _onRow$onClick === void 0 ? void 0 : _onRow$onClick.call(_onRow, e);
      onItem === null || onItem === void 0 ? void 0 : (_onItem = onItem(record, index)) === null || _onItem === void 0 ? void 0 : (_onItem$onClick = _onItem.onClick) === null || _onItem$onClick === void 0 ? void 0 : _onItem$onClick.call(_onItem, e);

      if (expandRowByClick) {
        onExpand(!expanded);
      }
    }
  }), /*#__PURE__*/React.createElement(_Skeleton, {
    avatar: true,
    title: false,
    loading: loading,
    active: true
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(className, "-header")
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(className, "-header-option")
  }, !!checkbox && /*#__PURE__*/React.createElement("div", {
    className: "".concat(className, "-checkbox")
  }, checkbox), Object.values(expandableConfig || {}).length > 0 && rowSupportExpand && renderExpandIcon({
    prefixCls: prefixCls,
    expandIcon: expandIcon,
    onExpand: onExpand,
    expanded: expanded,
    record: record
  })), (_ref4 = itemHeaderRender && (itemHeaderRender === null || itemHeaderRender === void 0 ? void 0 : itemHeaderRender(record, index, metaDom))) !== null && _ref4 !== void 0 ? _ref4 : metaDom), needExpanded && (content || expandedRowDom) && /*#__PURE__*/React.createElement("div", {
    className: "".concat(className, "-content")
  }, content, expandedRowRender && rowSupportExpand && /*#__PURE__*/React.createElement("div", {
    className: expandedRowClassName && expandedRowClassName(record, index, indentSize)
  }, expandedRowDom)))) : /*#__PURE__*/React.createElement(ProCard, _extends({
    bordered: true,
    loading: loading,
    hoverable: true
  }, cardProps, {
    title: cardTitleDom,
    subTitle: subTitle,
    extra: extraDom,
    actions: actionsDom,
    bodyStyle: _objectSpread({
      padding: 24
    }, cardProps.bodyStyle)
  }, onItem === null || onItem === void 0 ? void 0 : onItem(record, index)), /*#__PURE__*/React.createElement(_Skeleton, {
    avatar: true,
    title: false,
    loading: loading,
    active: true
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(className, "-header")
  }, itemTitleRender && (itemTitleRender === null || itemTitleRender === void 0 ? void 0 : itemTitleRender(record, index, titleDom)), content)));

  if (!cardProps) {
    return defaultDom;
  }

  return /*#__PURE__*/React.createElement("div", {
    className: classNames((_classNames6 = {}, _defineProperty(_classNames6, "".concat(className, "-card"), cardProps), _defineProperty(_classNames6, propsClassName, propsClassName !== defaultClassName), _classNames6)),
    style: style
  }, defaultDom);
}

export default ProListItem;
