// @ant-design/pro-skeleton/es/index.js
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
var _excluded = ["type"];
import React from 'react';
import DescriptionsPageSkeleton, { DescriptionsSkeleton, TableItemSkeleton, TableSkeleton } from './component/Descriptions';
import ListPageSkeleton, { ListSkeleton, ListSkeletonItem, ListToolbarSkeleton, PageHeaderSkeleton } from './component/List';
import ResultPageSkeleton from './component/Result';

var PageSkeleton = function PageSkeleton(_ref) {
  var _ref$type = _ref.type,
      type = _ref$type === void 0 ? 'list' : _ref$type,
      rest = _objectWithoutProperties(_ref, _excluded);

  if (type === 'result') {
    return /*#__PURE__*/React.createElement(ResultPageSkeleton, rest);
  }

  if (type === 'descriptions') {
    return /*#__PURE__*/React.createElement(DescriptionsPageSkeleton, rest);
  }

  return /*#__PURE__*/React.createElement(ListPageSkeleton, rest);
};

export { ListPageSkeleton, ListSkeleton, ListSkeletonItem, PageHeaderSkeleton, ListToolbarSkeleton, DescriptionsSkeleton, TableSkeleton, TableItemSkeleton };
export default PageSkeleton;
